# M3 T02

## Exercici 1
Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants. Justifica la teva elecció.

In [1]:
import pandas as pd

df = pd.read_csv('/Users/jordi/Documents/IT Academy/Data Science/Sprint 3/archive/DelayedFlights.csv' , index_col=0)
df = df.sample(n=200000, replace=False)
df

#Per començar importem la llibreria Pandas que és la que farem servir per treballar amb el dataset i els dataframes.
#El print el fem directament amb 'df' i no print('df') per fer-ho més visual.

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
4561611  2008      8           4          1   1626.0        1548   1743.0   
1067078  2008      2          15          5   1756.0        1700   2207.0   
1219578  2008      3          16          7   1956.0        1935   2119.0   
64555    2008      1          23          3   1604.0        1550   1624.0   
5585988  2008     10           2          4   2355.0        2316    549.0   
...       ...    ...         ...        ...      ...         ...      ...   
1760551  2008      3           3          1   2013.0        2005   2249.0   
1693563  2008      3          25          2   1328.0        1240   1543.0   
1600167  2008      3          12          3   1300.0        1040   1422.0   
640811   2008      2          13          3   1851.0        1800   2017.0   
3811258  2008      7          20          7   1302.0        1232   1427.0   

         CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
4561611        1708            EV       4320  ...    4.0     26.0          0   
1067078        2129            AA        744  ...    3.0     28.0          0   
1219578        2055            WN         75  ...    5.0     19.0          0   
64555          1615            WN       1676  ...    8.0     13.0          0   
5585988         506            UA        126  ...    6.0     15.0          0   
...             ...           ...        ...  ...    ...      ...        ...   
1760551        2252            CO       1449  ...    5.0      9.0          0   
1693563        1429            AA       1547  ...    7.0     31.0          0   
1600167        1215            MQ       4734  ...    8.0      7.0          0   
640811         1940            WN        470  ...    6.0     12.0          0   
3811258        1410            OO       5855  ...    4.0     12.0          0   

         CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
4561611                 N         0           0.0          0.0     35.0   
1067078                 N         0           0.0          0.0     20.0   
1219578                 N         0           2.0          0.0      3.0   
64555                   N         0           NaN          NaN      NaN   
5585988                 N         0           0.0          0.0      8.0   
...                   ...       ...           ...          ...      ...   
1760551                 N         0           NaN          NaN      NaN   
1693563                 N         0           0.0          0.0     74.0   
1600167                 N         0           0.0          0.0     14.0   
640811                  N         0          37.0          0.0      0.0   
3811258                 N         0           0.0          0.0      0.0   

         SecurityDelay  LateAircraftDelay  
4561611            0.0                0.0  
1067078            0.0               18.0  
1219578            0.0               19.0  
64555              NaN                NaN  
5585988            0.0               35.0  
...                ...                ...  
1760551            NaN                NaN  
1693563            0.0                0.0  
1600167            0.0              113.0  
640811             0.0                0.0  
3811258            0.0               17.0  

[200000 rows x 29 columns]

# Anàlisi del dataset i les seves variables.

Sobre el dataset, aquest realitza un seguiment de la puntualitat dels vols nacionals als EUA operats per grans companyies aèries. La informació disponible analitza la quantitat de vols a temps, amb retard, cancel·lats i desviats de vols del 2008.

Es realitzarà una anàlisi del dataset proporcionat, el qual conté fins a 1.936.758 vols interns diferents als EUA per 2008 i les seves causes de retard, desviament i cancel·lació. Les dades provenen de la Oficina d'Estadístiques de Transport (BTS) del Departament de Transport dels EUA (DOT). Per treballar sobre ell, agafarem un sample de 200000 dades.

Les variables que hi podem trobar són les següents:

* **Year**: L'any, el 2008.

* **Month**: Mesos, de l'1 al 12.
* **DayofMonth**: Dia del mes, de l'1 al 31.
* **DayofWeek**: Dia de la setmana, de l'1 (dilluns) al 7 (diumenge).
* **DepTime**: hora de sortida real (local, hhmm).
* **CRSDepTime**: hora de sortida programada (local, hhmm).
* **ArrTime**: hora d'arribada real (local, hhmm).
* **CRSArrTime**: hora d'arribada programada (local, hhmm).
* **UniqueCarrier**: codi únic de la companyia aèria.
* **FlightNum**: número de vol.
* **TailNum**: número de cua de l'avió: matrícula de l'aeronau, identificador únic de l'aeronau.
* **ActualElapsedTime**: Temps real transcorregut en minuts.
* **CRSElapsedTime**: Temps transcorregut en minuts.
* **AirTime**: en minuts.
* **ArrDelay**: retard en l'arribada, en minuts: Un vol es compta com "a temps" si es va realitzar amb menys de 15 minuts de retard respecte a l'hora programada que es mostra en els Sistemes de Reserves Computaritzades (CRS en anglés) de les aerolínies.
* **DepDelay**: retard en la sortida, en minuts.
* **Origin**: codi d'origen de l'aeroport IATA.
* **Dest**: codi de destí d'aeroport IATA.
* **Distance**: Distància en milles.
* **TaxiIn**: Temps en minuts quan l'aeronau te autorització per entrar a pista.
* **TaxiOut**: Temps en minuts quan l'aeronau surt de pista.
* **Cancelled**: Si el vol va ser cancel·lat o no: 0 = no, 1 = sí.
* **CancellationCode**: motiu de la cancel·lació (A = transportista, B = clima, C = NAS, D = seguretat).
* **Diverted**: Desviats: 1 = sí, 0 = no.
* **CarrierDelay**: retard en minuts de l'aeronau per culpa de l'aerolínia. Per exemple: repostatge de l'aeronau, neteja de la mateixa, avaria, etc.
* **WeatherDelay**: retard en minuts a conseqüència de condicions climatològiques adverses que poden transcòrrer abans de la sortida de l'aeroport orígen, durant el vol, o a l'aeroport destí. Per exemple: tempestes elèctriques, grans nevades, etc.
* **NASDelay**: en minuts: el retard que està sota el control del Sistema Nacional de l'Espai Aèri (NAS) pot incloure: condicions climàtiques no extremes, operacions aeroportuàries, volum de tràfic pesat, control de tràfic aèri, etc.
* **SecurityDelay**: en minuts: el retard de seguretat es produeix per l'evacuació d'una terminal o vestíbul, tornar a embarcar una aeronau a causa d'una violació de la seguretat, equips de control inoperatius i/o llargues files de més de 29 minuts als controls de seguretat de l'aeroport.
* **LateAircraftDelay**: en minuts: Retard en l'arribada a un aeroport a causa de l'arribada tardana de la mateixa aeronau a un aeroport anterior. L'efecte dominó d'un retard anterior als aeroports posteriors es denomina propagació del retard.




In [2]:
df = df.drop(['Year','FlightNum','TaxiIn', 'TaxiOut'], axis=1)
df

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
4561611      8           4          1   1626.0        1548   1743.0   
1067078      2          15          5   1756.0        1700   2207.0   
1219578      3          16          7   1956.0        1935   2119.0   
64555        1          23          3   1604.0        1550   1624.0   
5585988     10           2          4   2355.0        2316    549.0   
...        ...         ...        ...      ...         ...      ...   
1760551      3           3          1   2013.0        2005   2249.0   
1693563      3          25          2   1328.0        1240   1543.0   
1600167      3          12          3   1300.0        1040   1422.0   
640811       2          13          3   1851.0        1800   2017.0   
3811258      7          20          7   1302.0        1232   1427.0   

         CRSArrTime UniqueCarrier TailNum  ActualElapsedTime  ...  Dest  \
4561611        1708            EV  N842AS               77.0  ...   LEX   
1067078        2129            AA  N232AA              191.0  ...   LGA   
1219578        2055            WN  N669SW               83.0  ...   BNA   
64555          1615            WN  N249WN               80.0  ...   DEN   
5585988         506            UA  N670UA              234.0  ...   ORD   
...             ...           ...     ...                ...  ...   ...   
1760551        2252            CO  N39297              156.0  ...   EWR   
1693563        1429            AA  N207AA              195.0  ...   ORD   
1600167        1215            MQ  N709GB               82.0  ...   LGA   
640811         1940            WN  N774SW              146.0  ...   DEN   
3811258        1410            OO  N956SW               85.0  ...   FSD   

         Distance  Cancelled  CancellationCode Diverted CarrierDelay  \
4561611       303          0                 N        0          0.0   
1067078      1389          0                 N        0          0.0   
1219578       395          0                 N        0          2.0   
64555         358          0                 N        0          NaN   
5585988      1745          0                 N        0          0.0   
...           ...        ...               ...      ...          ...   
1760551      1024          0                 N        0          NaN   
1693563       867          0                 N        0          0.0   
1600167       431          0                 N        0          0.0   
640811        895          0                 N        0         37.0   
3811258       462          0                 N        0          0.0   

         WeatherDelay  NASDelay SecurityDelay  LateAircraftDelay  
4561611           0.0      35.0           0.0                0.0  
1067078           0.0      20.0           0.0               18.0  
1219578           0.0       3.0           0.0               19.0  
64555             NaN       NaN           NaN                NaN  
5585988           0.0       8.0           0.0               35.0  
...               ...       ...           ...                ...  
1760551           NaN       NaN           NaN                NaN  
1693563           0.0      74.0           0.0                0.0  
1600167           0.0      14.0           0.0              113.0  
640811            0.0       0.0           0.0                0.0  
3811258           0.0       0.0           0.0               17.0  

[200000 rows x 25 columns]

He decidit eliminar les columnes "Year", donat que l'any és sempre el mateix i ja el sabem (2008); "FlightNum" perquè el número de vol ens es irrellevant per l'anàlisi que volem fer, ja que tot i que podriem pensar que podriem trobar patrons que es repeteixin en determinades rutes, una mateixa ruta pot (i sol) estar coberta per diferents avions, llavors hi hauria un petit biaix. "TaxiIn i TaxiOut" ens aporta informació de les entrades i sortides de l'aeronau a pista.
A continuació explico perquè he decidit deixar la resta, anant columna per columna:

* **Month**: Ens pot servir per analitzar si hi ha algun patró que influeixi que aquell mes hi hagin més retards, com podrien ser condicions climatològiques adverses perllongades, per exemple.

* **DayofMonth**: Podriem fixar-nos amb detall si hi un dia en concret va ocórrer alguna situació extrema, ja sigui climàtica, de seguretat, etc.

* **DayofWeek**: Aquí podriem analitzar si hi ha algun patró que es repeteixi un dia concret de la setmana.

* **DepTime**: Necessari per controlar l'hora de sortida real de l'aeronau.

* **CRSDepTime**: Necessitem saber l'hora programada per comparar-la amb la real i poder calcul·lar la diferència real en minuts.

* El mateix passa amb les columnes d'arribada **ArrTime** i **CRSArrTime**.

* **UniqueCarrier**: Ens es útil per comprovar si hi ha aerolínies que acumulin més retards que d'altres, podent extreure la conclusió que la gestió te a veure amb els retards.

* **TailNum**: Podriem fer un cribatge i comprovar si una mateixa aeronau acumula més retards que d'altres, depenent del model, antiguitat, possibles avaries més sovint, etc.

* **ActualElapsedTime**: Necessari per saber el temps real transcorregut.

* **Distance**: Ens serveix per veure si hi ha influència en el radi de les rutes. Com per exemple podriem pensar que a major distància, més opcions de trobar climatologia variada i/o adversa.

* Les columnes **Cancelled** i **Diverted** per comprovar si hi ha hagut cancel·lació o desviació del vol.

* La columna **CancellationCode** per apreciar si el vol ha estat cancel·lat o no mitjançant el codi.

* **CarrierDelay**: Per veure el retard causat per l'aerolínia.

* **WeatherDelay**: Temps de retard causat per condicions climatològiques.

* **NASDelay**: Per comprovar el temps de retard causat per tancament d'espai aeri.

* **SecurityDelay**: Retard causat per problemes de seguretat.

* **LateAircraftDelay**: Ens és útil per saber si el retard ve causat perquè l'aeronau ja s'havia enlairat tard degut a un aterrament amb retard a l'aeroport orígen.

In [3]:
df.info()
#Exploració i tipus de dades que conté el dataframe.

<class 'pandas.core.frame.DataFrame'>
Int64Index: 200000 entries, 4561611 to 3811258
Data columns (total 25 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Month              200000 non-null  int64  
 1   DayofMonth         200000 non-null  int64  
 2   DayOfWeek          200000 non-null  int64  
 3   DepTime            200000 non-null  float64
 4   CRSDepTime         200000 non-null  int64  
 5   ArrTime            199231 non-null  float64
 6   CRSArrTime         200000 non-null  int64  
 7   UniqueCarrier      200000 non-null  object 
 8   TailNum            200000 non-null  object 
 9   ActualElapsedTime  199092 non-null  float64
 10  CRSElapsedTime     199985 non-null  float64
 11  AirTime            199092 non-null  float64
 12  ArrDelay           199092 non-null  float64
 13  DepDelay           200000 non-null  float64
 14  Origin             200000 non-null  object 
 15  Dest               200000 non-null  object 


In [4]:
df.head(20)
#Cridem els 20 primers valors.

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
4561611      8           4          1   1626.0        1548   1743.0   
1067078      2          15          5   1756.0        1700   2207.0   
1219578      3          16          7   1956.0        1935   2119.0   
64555        1          23          3   1604.0        1550   1624.0   
5585988     10           2          4   2355.0        2316    549.0   
1598987      3           4          2   1759.0        1725   1943.0   
17088        1           8          2   2134.0        2110      1.0   
136383       1          10          4    847.0         800   1040.0   
1777553      3           7          5   1613.0        1605   1821.0   
3221992      6          10          2   1051.0         955   1221.0   
3843634      7          11          5   1945.0        1859   2228.0   
6761117     12          28          7   2158.0        2126   2312.0   
753316       2          10          7   1744.0        1700   1830.0   
2352960      4           7          1   1110.0        1040   1432.0   
3886286      7          12          6   1731.0        1724   1940.0   
4274388      8          15          5   1944.0        1900   2040.0   
4235361      8           3          7   1912.0        1900   2050.0   
1468888      3          25          2   2157.0        2025     34.0   
4386872      8           1          5   1255.0        1245   1410.0   
3692727      7          29          2    617.0         610    738.0   

         CRSArrTime UniqueCarrier TailNum  ActualElapsedTime  ...  Dest  \
4561611        1708            EV  N842AS               77.0  ...   LEX   
1067078        2129            AA  N232AA              191.0  ...   LGA   
1219578        2055            WN  N669SW               83.0  ...   BNA   
64555          1615            WN  N249WN               80.0  ...   DEN   
5585988         506            UA  N670UA              234.0  ...   ORD   
1598987        1910            MQ  N724AE              104.0  ...   RDU   
17088          2350            WN  N286WN               87.0  ...   IAD   
136383          906            YV  N437YV              113.0  ...   PHX   
1777553        1752            DL  N915DL              128.0  ...   ATL   
3221992        1130            UA  N675UA              150.0  ...   DEN   
3843634        2148            UA  N349UA              103.0  ...   MDT   
6761117        2243            EV  N760EV               74.0  ...   DAB   
753316         1744            OH  N973CA               46.0  ...   DAY   
2352960        1339            CO  N62631              262.0  ...   IAH   
3886286        1912            US  N677AW              309.0  ...   PHX   
4274388        2000            WN  N302SW               56.0  ...   HOU   
4235361        2050            WN  N370SW               98.0  ...   SJC   
1468888        2325            US  N947UW               97.0  ...   PHL   
4386872        1412            OH  N591SW               75.0  ...   BOS   
3692727         730            WN  N514SW               81.0  ...   HOU   

         Distance  Cancelled  CancellationCode Diverted CarrierDelay  \
4561611       303          0                 N        0          0.0   
1067078      1389          0                 N        0          0.0   
1219578       395          0                 N        0          2.0   
64555         358          0                 N        0          NaN   
5585988      1745          0                 N        0          0.0   
1598987       532          0                 N        0          0.0   
17088         577          0                 N        0          NaN   
136383        160          0                 N        0          0.0   
1777553       526          0                 N        0          8.0   
3221992       888          0                 N        0         42.0   
3843634       594          0                 N        0         40.0   
6761117       366          0                 N        0          0.0   
75331

# Exercici 2
Fes un informe complet del dataset:

* Resumeix estadísticament el dataset i les columnes d’interès. Fes una anàlisi estadístic del que consideris rellevant.

* Troba quantes dades faltants hi ha per columna.
* Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).
* Fes una taula de les aerolínies amb més endarreriments acumulats.
* Quins són els vols més llargs? I els més endarrerits? Busca les rutes més llargues i les que acumulen més retards.

Aporta allò que consideris rellevant.


In [5]:
df.describe().transpose()
#Resum estadístic del dataset amb el mètode .transpose() per facilitar visualització del total de columnes.

count         mean         std   min     25%     50%  \
Month              200000.0     6.123975    3.486387   1.0     3.0     6.0   
DayofMonth         200000.0    15.738615    8.776717   1.0     8.0    16.0   
DayOfWeek          200000.0     3.984685    1.996189   1.0     2.0     4.0   
DepTime            200000.0  1518.276515  450.956095   1.0  1202.0  1546.0   
CRSDepTime         200000.0  1467.683155  424.931709   1.0  1135.0  1510.0   
ArrTime            199231.0  1610.914617  547.401998   1.0  1316.0  1715.0   
CRSArrTime         200000.0  1633.959875  464.683435   0.0  1325.0  1705.0   
ActualElapsedTime  199092.0   133.217337   72.034078  15.0    80.0   116.0   
CRSElapsedTime     199985.0   134.228177   71.296658   1.0    81.0   116.0   
AirTime            199092.0   108.175718   68.620698   1.0    58.0    90.0   
ArrDelay           199092.0    42.222334   56.738951 -60.0     9.0    25.0   
DepDelay           200000.0    43.218360   53.310973   6.0    12.0    24.0   
Distance           200000.0   764.624265  573.426085  31.0   338.0   606.0   
Cancelled          200000.0     0.000395    0.019871   0.0     0.0     0.0   
Diverted           200000.0     0.004145    0.064248   0.0     0.0     0.0   
CarrierDelay       129111.0    19.070436   43.124391   0.0     0.0     2.0   
WeatherDelay       129111.0     3.693752   21.272981   0.0     0.0     0.0   
NASDelay           129111.0    15.218262   34.599726   0.0     0.0     2.0   
SecurityDelay      129111.0     0.082952    1.724112   0.0     0.0     0.0   
LateAircraftDelay  129111.0    25.129702   41.972394   0.0     0.0     8.0   

                      75%     max  
Month                 9.0    12.0  
DayofMonth           23.0    31.0  
DayOfWeek             6.0     7.0  
DepTime            1900.0  2400.0  
CRSDepTime         1815.0  2359.0  
ArrTime            2031.0  2400.0  
CRSArrTime         2015.0  2400.0  
ActualElapsedTime   165.0  1114.0  
CRSElapsedTime      165.0   660.0  
AirTime             137.0  1091.0  
ArrDelay             56.0  1395.0  
DepDelay             53.0  1401.0  
Distance            998.0  4962.0  
Cancelled             0.0     1.0  
Diverted              0.0     1.0  
CarrierDelay         21.0  1395.0  
WeatherDelay          0.0   907.0  
NASDelay             15.0  1195.0  
SecurityDelay         0.0   136.0  
LateAircraftDelay    33.0  1236.0

* Al resum estadístic podem trobar dades interessants com ara que al menys el 50% dels vols han tingut una distància de 606 milles, el més llarg 4962 milles, o el més curt 31.

* També podem extreure la conclusió que en relació als retards, el més gran ha sigut per culpa de l'aerolíniea, el segon per arribada tardana de l'aeronau i el tercer per tancament de l'espai aeri. . També podem trobar dades dubtoses com ara que el temps més curt que una aeronau hagi estat enlairada fos 1 minut.

* De mitjana els retards a l'arribada son de 42,22 minuts.

* De mijtana els vols duren uns 108,18 minuts enlairats. 

### - Troba quantes dades faltants hi ha per columna.

In [6]:
df.isna().sum() #Per trobar dades faltants per columna.

Month                    0
DayofMonth               0
DayOfWeek                0
DepTime                  0
CRSDepTime               0
ArrTime                769
CRSArrTime               0
UniqueCarrier            0
TailNum                  0
ActualElapsedTime      908
CRSElapsedTime          15
AirTime                908
ArrDelay               908
DepDelay                 0
Origin                   0
Dest                     0
Distance                 0
Cancelled                0
CancellationCode         0
Diverted                 0
CarrierDelay         70889
WeatherDelay         70889
NASDelay             70889
SecurityDelay        70889
LateAircraftDelay    70889
dtype: int64

Observem que 10 columnes del dataset a analitzar presenten files amb absència de dades. Això és degut a que si un vol es desviat o cancel·lat pot no presentar dades a les altres columnes d'interès.

### - Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).

In [7]:
#Fem selecció de les dades que ens interessen:
df.loc[df['ArrDelay'] <= 15, 'Status'] = "On time"
df.loc[df['ArrDelay'] > 15, 'Status'] = "Delayed"
df.loc[df['Diverted'] == 1,  'Status'] = "Diverted"
df.loc[df['Cancelled'] == 1, 'Status'] = "Cancelled"

#Eliminem els no desitjats
df = df.drop(['Diverted'], axis=1)
df = df.drop(['Cancelled'], axis=1)

#Filtrem i mostrem els que porten retard:
df2 = df[df.Status != "On time"]

Segons el Sistema de Reserves Computaritzades (CRS) de les aerolínees, un vol arriba "a temps" si opera en menys de 15 minuts després de l'hora programada, llavors en base a això apliquem un condicional per saber l'status del vol. 

In [8]:
import warnings
warnings.filterwarnings('ignore')

df2['Distance_km'] = df2['Distance'] * 1.60934 #Convertim milles a km.
df2['Vmean'] = df2['Distance_km'] / (df2['AirTime']/60) #Creem una nova columna amb la velocitat mitjana de l'aparell enlairat, en km/h.

In [9]:
df2.head(20)

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
4561611      8           4          1   1626.0        1548   1743.0   
1067078      2          15          5   1756.0        1700   2207.0   
1219578      3          16          7   1956.0        1935   2119.0   
5585988     10           2          4   2355.0        2316    549.0   
1598987      3           4          2   1759.0        1725   1943.0   
136383       1          10          4    847.0         800   1040.0   
1777553      3           7          5   1613.0        1605   1821.0   
3221992      6          10          2   1051.0         955   1221.0   
3843634      7          11          5   1945.0        1859   2228.0   
6761117     12          28          7   2158.0        2126   2312.0   
753316       2          10          7   1744.0        1700   1830.0   
2352960      4           7          1   1110.0        1040   1432.0   
3886286      7          12          6   1731.0        1724   1940.0   
4274388      8          15          5   1944.0        1900   2040.0   
1468888      3          25          2   2157.0        2025     34.0   
6232361     11          14          5   2324.0        2122     53.0   
407856       1          11          5    843.0         815   1041.0   
6591562     12          18          4   1418.0        1237   1545.0   
5193298      9          18          4   1013.0         950   1113.0   
2559843      5           2          5   1805.0        1754   1947.0   

         CRSArrTime UniqueCarrier TailNum  ActualElapsedTime  ...  Distance  \
4561611        1708            EV  N842AS               77.0  ...       303   
1067078        2129            AA  N232AA              191.0  ...      1389   
1219578        2055            WN  N669SW               83.0  ...       395   
5585988         506            UA  N670UA              234.0  ...      1745   
1598987        1910            MQ  N724AE              104.0  ...       532   
136383          906            YV  N437YV              113.0  ...       160   
1777553        1752            DL  N915DL              128.0  ...       526   
3221992        1130            UA  N675UA              150.0  ...       888   
3843634        2148            UA  N349UA              103.0  ...       594   
6761117        2243            EV  N760EV               74.0  ...       366   
753316         1744            OH  N973CA               46.0  ...        64   
2352960        1339            CO  N62631              262.0  ...      1416   
3886286        1912            US  N677AW              309.0  ...      1972   
4274388        2000            WN  N302SW               56.0  ...       239   
1468888        2325            US  N947UW               97.0  ...       678   
6232361        2254            EV  N917EV               89.0  ...       508   
407856         1015            MQ  N686AE               58.0  ...       177   
6591562        1349            YV  N7264V               87.0  ...       288   
5193298        1045            MQ  N827AE               60.0  ...       190   
2559843        1927            OO  N773SK              162.0  ...       960   

         CancellationCode  CarrierDelay  WeatherDelay NASDelay SecurityDelay  \
4561611                 N           0.0           0.0     35.0           0.0   
1067078                 N           0.0           0.0     20.0           0.0   
1219578                 N           2.0           0.0      3.0           0.0   
5585988                 N           0.0           0.0      8.0           0.0   
1598987                 N           0.0           1.0      0.0           0.0   
136383                  N           0.0           0.0     94.0           0.0   
1777553                 N           8.0           0.0     21.0           0.0   
3221992                 N          42.0           0.0      0.0           0.0   
3843634                 N          40.0           0.0      0.0           0.0   
6761117                 N           0.0           0.0    

### Fes una taula de les aerolínies amb més endarreriments acumulats.

In [10]:
df2.UniqueCarrier.value_counts().head(10)
#Mostrem les 10 aerolínees que presenten més retards.

WN    20514
AA    13482
MQ     9943
UA     9585
OO     9115
DL     7442
XE     7359
US     6135
CO     5919
EV     5642
Name: UniqueCarrier, dtype: int64

In [11]:
df2.UniqueCarrier.value_counts(ascending=True).head(10)
#Aquí mostrem les aerolínees amb menys retard, sent la primera de la llista la que menys retards presenta, i així succesivament.
#He optat per fer-ho així i no amb un .tail() per fer-ho més visual.

AQ      30
HA     402
F9    1603
AS    2463
9E    3673
B6    3946
OH    3969
FL    4711
YV    5176
NW    5563
Name: UniqueCarrier, dtype: int64

D'aquest petit anàlisi, si comprovem el codi IATA de les aerolínies, trobem que la que presenta més retards és Southwest Airlines (WN) una aerolínia de baix cost amb gran presència al llarg de tot el territori dels EUA, amb base a diferents ciutats on les condicions climatològiques solen ser adverses: Chicago molt de vent; Fort Lauderdale a Florida on pateixen temporades d'huracans, etc…
Per una altra banda, la que menys, Aloha Airlines (AQ) és una companyia que opera pocs vols, el qual ja fa incrementar la possibilitat de tenir menys retards pel pur volum de passatgers i disponibilitat d'aeronaus, així com s'afavoreix del bon clima d'on te la base operacional, situada a Hawaii, on gaudeixen de bon temps la gran majoria de l'any.

### Quins són els vols més llargs? I els més endarrerits? Busca les rutes més llargues i les que acumulen més retards.

In [12]:
ruta_mes_llarga = df2[df2.Distance == df2.Distance.max()]
ruta_mes_llarga[['Origin','Dest','Distance']].drop_duplicates()

Origin Dest  Distance
6432273    EWR  HNL      4962
1742517    HNL  EWR      4962

La ruta més llarga és la que va de Honolulu, Hawaii (HNL) a Newark, Nova Jersey (EWR), amb un total de 4.962 milles o 7.985 kilòmetres.

In [13]:
df2[['Origin','Dest','Distance']].sort_values('Distance',ascending=False).drop_duplicates().head(10)
#Rutes més llargues.

Origin Dest  Distance
2962775    EWR  HNL      4962
5357897    HNL  EWR      4962
1164647    ATL  HNL      4502
3308476    HNL  ATL      4502
6660952    HNL  ORD      4243
4097746    ORD  HNL      4243
5063550    KOA  ORD      4213
6684808    ORD  OGG      4184
4679392    MSP  HNL      3972
2828211    HNL  MSP      3972

In [14]:
delay = df2[df2.ArrDelay == df2.ArrDelay.max()]
delay[['Origin','Dest','ArrDelay']]

Origin Dest  ArrDelay
2234143    SJU  DTW    1395.0

El vol més retrassat acumula 1395 minuts o 23,25 hores de San Juan, Puerto Rico (SJU) a Dallas, Texas (DFW).

In [15]:
df2[['Origin','Dest','ArrDelay']].sort_values('ArrDelay',ascending=False).drop_duplicates().head(10)
#Vols més endarrerits.

Origin Dest  ArrDelay
2234143    SJU  DTW    1395.0
6308375    MEM  MIA    1253.0
1030111    EGE  MSP    1249.0
1018915    MCO  IND    1169.0
1621875    FCA  MSP    1166.0
2838200    ALB  DTW    1155.0
997272     SHV  ORD    1142.0
5273496    LIH  LAX    1113.0
3445213    GFK  MSP    1085.0
997219     AZO  ORD    1064.0

### Aporta allò que consideris rellevant.

In [22]:
df.Status.value_counts()

Delayed      125764
On time       73328
Diverted        829
Cancelled        79
Name: Status, dtype: int64

Podem apreciar de la mostra de 200.000 vols analitzada, que 125764 van patir algun retard, 73328 van arribar a temps, 829 van ser desviats, i 79 cancel·lats.

In [30]:
df2.loc[df2.Status == 'Delayed', 'DayOfWeek'].value_counts()

5    21640
4    18834
7    18772
1    18750
2    17284
3    16892
6    13592
Name: DayOfWeek, dtype: int64

Aquí he mostrat el número total de retards per dia, en ordre acumulat de més a menys retards. El día que presenta una incidència més elevada de retards és el divendres, i el que menys, el dissabte.

# Exercici 3

Exporta el dataset net i amb les noves columnes a Excel.

In [18]:
import openpyxl as open
df2.to_excel('dataset.xlsx', index = False)